internal/controller/tlsconfigutil/tls_config_util.go

```diff
@@ -0,0 +1,232 @@
+		return nil, nil, fmt.Errorf("%s is invalid: %s", field, ErrNoCertificates)
+	}
+
+	return ca, bundleBytes, nil
+}
+
+// ValidateTLSConfig reads ca bundle in the tlsSpec, supplied either inline using the CertificateAuthorityDate
+// or as a reference to a kubernetes secret or configmap using the CertificateAuthorityDataSource, and returns
+// a condition of type TLSConfigurationValid based on the validity of the ca bundle,
+// a pem encoded ca bundle
+// a X509 cert pool with the ca bundle
+// any error encountered.
+// TODO: it should suffice that this function returns a TLSConfigurationValid condition, and perhaps we could skip
+// returning the error. This can be done once all controllers are able to use this function.
+func ValidateTLSConfig(
+	tlsSpec *TLSSpec,
+	conditionPrefix string,
+	namespace string,
+	secretInformer v1.SecretInformer,
+	configMapInformer v1.ConfigMapInformer,
+) (*v12.Condition, []byte, *x509.CertPool, error) {
+	// try to build a x509 cert pool using the ca data specified in the tlsSpec.
+	certPool, bundle, err := getCertPool(tlsSpec, conditionPrefix, namespace, secretInformer, configMapInformer)
+	if err != nil {
+		// an error encountered during building a certpool using the ca data from the tlsSpec results in an invalid
+		// TLS condition.
+		return invalidTLSCondition(err.Error()), nil, nil, err
+	}
+	// for us, an empty or nil ca bundle read is results in a valid TLS condition, but we do want to convey that
+	// no ca data was supplied.
+	if bundle == nil {
+		return validTLSCondition(fmt.Sprintf("%s is valid: %s", conditionPrefix, noTLSConfigurationMessage)), bundle, certPool, err
+	}
+
+	return validTLSCondition(fmt.Sprintf("%s is valid: %s", conditionPrefix, loadedTLSConfigurationMessage)), bundle, certPool, err
+}
+
+func readCABundleFromSource(source *caBundleSource, namespace string, secretInformer v1.SecretInformer, configMapInformer v1.ConfigMapInformer) (string, error) {
+	switch source.Kind {
+	case "Secret":
+		return readCABundleFromK8sSecret(namespace, source.Name, source.Key, secretInformer)
+	case "ConfigMap":
+		return readCABundleFromK8sConfigMap(namespace, source.Name, source.Key, configMapInformer)
+	default:
+		return "", fmt.Errorf("unsupported CA bundle source: %s", source.Kind)
+	}
+}
+
+func readCABundleFromK8sSecret(namespace string, name string, key string, secretInformer v1.SecretInformer) (string, error) {
+	s, err := secretInformer.Lister().Secrets(namespace).Get(name)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to get secret %s/%s", namespace, name)
+	}
+	// ca bundle in the secret is expected to exist in a specific key, if that key does not exist, then it is an error
+	if val, exists := s.Data[key]; exists {
+		return string(val), nil
+	}
+	return "", fmt.Errorf("key %s not found in secret %s/%s", key, namespace, name)
+}
+
+func readCABundleFromK8sConfigMap(namespace string, name string, key string, configMapInformer v1.ConfigMapInformer) (string, error) {
```

Comment:
This should validate the secret type and return an error if the secret type is not `Opaque` or `kubernetes.io/tls`